### PR TITLE
ast: Keep expansion status for out-of-line module items

### PR DESCRIPTION
**Pull request by petrochenkov**
I.e. whether a module `mod foo;` is already loaded from a file or not.
This is a pre-requisite to correctly treating inner attributes on such modules (https://github.com/rust-lang/rust/issues/81661).

With this change AST structures for `mod` items diverge even more for AST structure for the crate root, which previously used `ast::Mod`.
Therefore this PR removes `ast::Mod` from `ast::Crate` in the first commit, these two things are sufficiently different from each other, at least at syntactic level.
Customization points for visiting a "`mod` item or crate root" were also removed from AST visitors (`fn visit_mod`).
`ast::Mod` itself was refactored away in the second commit in favor of `ItemKind::Mod(Unsafe, ModKind)`.

### CODE REVIEW

File: compiler/rustc_ast/src/ast.rs [hunk 1]
```diff
@@ -2299,21 +2299,17 @@ impl FnRetTy {
...
-    pub items: Vec<P<Item>>,
-    /// `true` for `mod foo { .. }`; `false` for `mod foo;`.
-    pub inline: bool,
+pub enum ModKind {
+    /// Module with inlined definition `mod foo { ... }`.
+    /// The inner span is from the first token past `{` to the last token until `}`.
+    Inlined(Vec<P<Item>>, Span),
+    /// Module with definition outlined to a separate file `mod foo;` and already loaded from it.
+    /// The inner span ranges from the first to the last token in the loaded file.
+    OutlinedLoaded(Vec<P<Item>>, Span),
```
Comment (petrochenkov) [inline compiler/rustc_ast/src/ast.rs:2310]: Alternative 1: rename this to just `Outlined` for brevity, it's hard to mix up loaded and unloaded modules due to different fields.

Alternative 2: two-variant enum
```rust
enum ModKind {
    Loaded(Vec<P<Item>>, Span, IsInline),
    Unloaded,
}
```
This should be more convenient because `Inlined` and `OutlinedLoaded` are almost always treated in the same way (to obtain items or inner span from them, see the second commit). I'm not sure whether it would be more clear though.

File: src/test/ui/issues/issue-49040.stderr [hunk 2]
```diff
@@ -7,10 +7,8 @@ LL | #![allow(unused_variables)];
 error[E0601]: `main` function not found in crate `issue_49040`
   --> $DIR/issue-49040.rs:1:1
    |
-LL | / #![allow(unused_variables)];
-LL | |
-LL | | fn foo() {}
-   | |__^ consider adding a `main` function to `$DIR/issue-49040.rs`
+LL | #![allow(unused_variables)];
+   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ consider adding a `main` function to `$DIR/issue-49040.rs`
```
Comment (Aaron1011) [inline src/test/ui/issues/issue-49040.stderr:11]: Is this change intentional? There's not really a great span for this error, so I suppose an attribute is as good as any.

Comment (petrochenkov) [reply src/test/ui/issues/issue-49040.stderr:11]: Yes, there's a slight change in the crate span in case of parsing errors due to changes in `fn parse_crate_mod`.
Previously it went one token ahead into the unparsed part of the crate (see how `^` points to `fn`), now it covers the parsed part consistently with non-root modules.

File: compiler/rustc_ast/src/mut_visit.rs [hunk 3]
```diff
@@ -1014,19 +1010,18 @@ pub fn noop_visit_crate<T: MutVisitor>(krate: &mut Crate, vis: &mut T) {
             id: DUMMY_NODE_ID,
             vis: item_vis,
             span,
-            kind: ItemKind::Mod(module),
+            kind: ItemKind::Mod(Mod { inner: span, unsafety: Unsafe::No, items, inline: true }),
```
Comment (Aaron1011) [inline compiler/rustc_ast/src/mut_visit.rs]: Constructing this synthetic `Mod` seems kind of unfortunate, but it looks like it's necessary for the `flat_map_item` handling below.

File: compiler/rustc_ast_lowering/src/item.rs [hunk 4]
```diff
@@ -34,25 +34,6 @@ impl ItemLowerer<'_, '_, '_> {
 }
 
 impl<'a> Visitor<'a> for ItemLowerer<'a, '_, '_> {
-    fn visit_mod(&mut self, m: &'a Mod, _s: Span, _attrs: &[Attribute], n: NodeId) {
-        let def_id = self.lctx.lower_node_id(n).expect_owner();
-
-        self.lctx.modules.insert(
```
Comment (Aaron1011) [inline compiler/rustc_ast_lowering/src/item.rs:40]: Could this be moved to the `ItemKind::Mod(..) => ` match arm below? That would allow you to avoid using `lctx.modules.entry` in some of the other code.

File: compiler/rustc_ast/src/mut_visit.rs (hunk 3 above)
Comment (petrochenkov) [reply compiler/rustc_ast/src/mut_visit.rs]: I don't like this either and left a FIXME above.
I'll try to investigate removing this some time later, together with similar code in `fn expand_crate`.

File: compiler/rustc_ast/src/ast.rs (hunk 1 above)
Comment (Aaron1011) [reply compiler/rustc_ast/src/ast.rs:2310]: I have a slight preference for the two-variant enum, but I'll leave the final decision up to you.

File: compiler/rustc_ast_lowering/src/item.rs (hunk 4 above)
Comment (petrochenkov) [reply compiler/rustc_ast_lowering/src/item.rs:40]: I think `entry` is preferable, why pre-inserting empty values explicitly if `entry` can do everything for you automatically.

File: src/tools/clippy/clippy_lints/src/utils/ast_utils.rs [hunk 5]
```diff
@@ -241,9 +241,14 @@ pub fn eq_item_kind(l: &ItemKind, r: &ItemKind) -> bool {
...
                 && both(lb, rb, |l, r| eq_block(l, r))
         }
-        (Mod(l), Mod(r)) => {
-            l.inline == r.inline && over(&l.items, &r.items, |l, r| eq_item(l, r, eq_item_kind))
-        }
+        (Mod(lu, lmk), Mod(ru, rmk)) => lu == ru && match (lmk, rmk) {
+            (ModKind::Inlined(litems, _), ModKind::Inlined(ritems, _)) =>
+                over(litems, ritems, |l, r| eq_item(l, r, eq_item_kind)),
+            (ModKind::OutlinedLoaded(litems, _), ModKind::OutlinedLoaded(ritems, _)) =>
+                over(litems, ritems, |l, r| eq_item(l, r, eq_item_kind)),
```
Comment (Aaron1011) [inline src/tools/clippy/clippy_lints/src/utils/ast_utils.rs:248]: It looks like this arm can be combined with the previous one

File: compiler/rustc_ast_lowering/src/item.rs (hunk 4 above)
Comment (Aaron1011) [reply compiler/rustc_ast_lowering/src/item.rs:40]: I thought that the `get_mut().unwrap()` might have been a check to ensure that we had already processed the module in some way. If that's not the case, then I think your change is fine.

File: compiler/rustc_ast/src/ast.rs (hunk 1 above)
Comment (petrochenkov) [reply compiler/rustc_ast/src/ast.rs]: Switched to the alternative 2.

File: compiler/rustc_ast_pretty/src/pprust/state.rs [hunk 6]
```diff
@@ -1138,23 +1129,29 @@ impl<'a> State<'a> {
...
-                    self.nbsp();
-                    self.bopen();
-                    self.print_mod(_mod, &item.attrs);
-                    self.bclose(item.span);
-                } else {
-                    self.s.word(";");
-                    self.end(); // end inner head-block
-                    self.end(); // end outer head-block
+                match mod_kind {
+                    ModKind::Loaded(items, ..) => {
```
Comment (Aaron1011) [inline compiler/rustc_ast_pretty/src/pprust/state.rs:1141]: It looks like this will always pretty-print the contents of outline modules - is that intentional?

**Comment by Aaron1011** [comment]
r=me assuming that the pretty-printer change is intentional.

File: compiler/rustc_ast_pretty/src/pprust/state.rs (hunk 6 above)
Comment (petrochenkov) [reply compiler/rustc_ast_pretty/src/pprust/state.rs:1141]: For `--pretty=normal` and `--pretty=expanded` this is a replacement of the removed `self.is_expanded` flag.

For synthesizing tokens for inner macro attributes on modules (which are broken until https://github.com/rust-lang/rust/issues/81661 is fixed) I think it also makes sense to tokenize the whole module rather than just the `mod foo;` header.
